File: ARGContinuousScroll/ARGContinuousScroll/ARGScrollViewDelegateProxy.m
//
//  ARGScrollViewDelegateProxy.m
//  ARGView
//
//  Created by Sergei Polshcha on 17.10.2020.
//  Copyright © 2020 Argentum. All rights reserved.
//

#import "ARGScrollViewDelegateProxy.h"

@interface ARGScrollViewDelegateProxy()

@property (nonatomic) NSMutableArray *delegates;

@end

@implementation ARGScrollViewDelegateProxy

- (void)addDelegate:(id<UIScrollViewDelegate>)delegate {
    if (!_delegates) {
        _delegates = [NSMutableArray new];
    }
    
    [_delegates addObject:delegate];
}

- (void)removeDelegate:(id<UIScrollViewDelegate>)delegate {
    [_delegates removeObject:delegate];
}

- (void)forwardInvocation:(NSInvocation *)anInvocation {
    SEL selector = [anInvocation selector];
    BOOL forwarded = NO;
    
    for (id <UIScrollViewDelegate> delegate in _delegates) {
        if ([delegate respondsToSelector:selector]) {
            [anInvocation invokeWithTarget:delegate];
            forwarded = YES;
        }
    }
    
    if (!forwarded) {
        [super forwardInvocation:anInvocation];
    }
}

@end

File: ARGContinuousScroll/ARGContinuousScroll/ARGScrollViewDelegateProxy.swift
//
//  ARGScrollViewDelegateProxy.swift
//  ARGContinuousScroll
//
//  Created by Sergei Polshcha on 21.10.2020.
//

import UIKit

public class ARGScrollViewDelegateProxy: NSObject, UIScrollViewDelegate {
    
    var delegates: [NSObject & UIScrollViewDelegate] = [NSObject & UIScrollViewDelegate]()
    
    public func addDelegate(_ delegate: NSObject & UIScrollViewDelegate) {
        delegates.append(delegate)
    }
    
    public func removeDelegate(_ delegate: UIScrollViewDelegate) {
        delegates = delegates.filter {!$0.isEqual(delegate)}
    }
    
    public override func forwardingTarget(for aSelector: Selector!) -> Any? {
        for delegate in delegates {
            if delegate.responds(to: aSelector) {
                return delegate
            }
        }
        
        return super.forwardingTarget(for: aSelector)
    }
    
    public override func responds(to aSelector: Selector!) -> Bool {
        for delegate in delegates {
            if delegate.responds(to: aSelector) {
                return true
            }
        }
        
        return false
    }
    
}
